src/impl/tpls/KokkosBlas3_gemm_tpl_spec_decl.hpp

```diff
@@ -335,6 +335,127 @@ KOKKOSBLAS3_CGEMM_BLAS( Kokkos::LayoutRight, Kokkos::LayoutRight, Kokkos::Layout
 namespace KokkosBlas {
 namespace Impl {
 
+
+// DotBasedGEMM implements the optimization for C = beta*C + alpha*A^TB 
+// with A and B matrices both being tall and skinny. C matrix is assumably 
+// small, so, each entry of C is computed by performing the dot product of 
+// respective columns of A and B matrices. Note that the dot products are
+// performed on very long vectors, so, each dot product is distributed among
+// numDivPerDot teams.     
+
+struct TagInit{};   // The tag for the initialization parallel_for 
+struct TagMult{};   // The tag for the multiplication parallel_for 
+template<class ExecSpace, class AV, class BV, class CV>
+struct DotBasedGEMM{
+
+  AV A;
+  BV B;
+  CV C;
+
+  typedef typename AV::non_const_value_type scalar_A;
+  typedef typename AV::size_type size_A;
+  typedef typename CV::non_const_value_type scalar_C;
+  typedef typename CV::size_type size_C;
+
+  scalar_A alpha;
+  scalar_C beta;
+
+  // The following types (especially dotSize) could have simply been int,
+  // e.g., see lines 488-490.
+  size_C numCrows;           
+  size_C numCcols;
+
+  size_C numDivPerDot;   // number of teams collectively performing a dot product
+  size_C numTeams;       // total number of teams
+  
+  size_A dotSize;        // the length of the vectors in the dot products
+  size_A chunkSize;      // the local length of each team's share on the dot product  
+  
+
+  DotBasedGEMM(const scalar_A& alpha_, const AV& A_, const BV& B_, const scalar_C& beta_, const CV& C_):A(A_),B(B_),C(C_),alpha(alpha_),beta(beta_) 
+  {
+    numCrows = C.extent(0);
```

Comment:
Why not initialize these in the initializer list?

Reply:
I added them to the initializer list.